[Exam][Code-Editor] Minor bugfixes

anditurdiu opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context
- **Save, Submit, Refresh files not available in exam mode after being offline:**
See #1797
- **Programming Excercise topbar indicator misbehaviour**
See #1819 

### Description
When the client is disconnected from the internet before or during saving, no error is thrown and the action buttons remain locked. In order to not lock the student action buttons in case the save operation does not return, we set a timeout of 8 seconds. If the timeout is reached and the editor is still in 'saving' mode, we set the mode to unsaved changes.

The topbar was only updated on file creation/deletion but not on file content changes. To fix this we handle the file content changes and update the topbar accordingly. 

### Steps for Testing
- **Save, Submit, Refresh files not available in exam mode after being offline:**

1. Create an exam with one programming exercise (make sure the code editor is enabled) and one other exercise
2. Participate in the exam and navigate to the programming exercise
3. Disconnect and wait until Artemis says you are disconnected
4. Make some changes, e.g. type some text into a file
5. Hit save or wait for the automatic save to trigger, this will lock the Save, Submit and Refresh files buttons
6. It should unlock once 8 seconds have passed

- **Programming Excercise topbar indicator misbehaviour**

1. Start an exam with a programming exercise
2. Navigate to the programming exercise
3. Change something in the online editor (no save or submission)
4. See that the top bar indicator updates

alexmardale commented:
Worked as expected!

Probably not related to this PR, but I just noticed one thing, the files don't seem to be loaded until you click on them. I started the exam, turned off the internet connection and tried to open the files, but they appeared as empty. They were only loaded after I turned on the internet connection and clicked on them again.
![2020-07-06-20-57-14](https://user-images.githubusercontent.com/39967206/86630145-bf2f4080-bfcc-11ea-8030-dd1d96804b0d.gif)

krusche commented:
Yes, this is an online code editor, so you need to be online to display the files. Offline support here is really limited and we just want to make sure that it does not cause any problems.
@anditurdiu we should still show an error message in such a case, could you add one as a red alert?

anditurdiu commented:
The timeout function now emits a timeout error which is displayed to the user. I also added the translations. 
I had to remove the jhi-alert on the exam editor container as it was displaying the error messages twice, once itself, once through its child views.

krusche commented on src/main/webapp/i18n/de/editor.json:
```diff
@@ -85,7 +85,8 @@
                 "repositoryInConflict": "Dein Repository befindet sich in einem Konfliktzustand.",
-                "notAllowedExam": "Du kannst (nicht mehr) abgeben"
+                "notAllowedExam": "Du kannst (nicht mehr) abgeben",
+                "connectionTimeout": "Zeitlimit überschritten. Speichern ist fehlgeschlagen."
```
it's not really clear what you mean with "Zeitlimit" this is too technical. I would rather translate it to user language.

krusche replied:
```suggestion
                "connectionTimeout": "Speichern ist fehlgeschlagen. Bitte stelle eine gute Internetverbindung sicher und versuche es nochmal."
```

krusche commented on src/main/webapp/i18n/en/editor.json:
```diff
@@ -89,7 +89,8 @@
                 "repositoryInConflict": "Your repository has entered a conflict state.",
-                "notAllowedExam": "You must not submit (anymore)"
+                "notAllowedExam": "You must not submit (anymore)",
+                "connectionTimeout": "Connection timeout. Saving failed."
```
```suggestion
                "connectionTimeout": ".Saving failed. Please make sure you have a stable internet connection and try again."
```

jpbernius commented:
<img width="909" alt="image" src="https://user-images.githubusercontent.com/6382716/86745873-f2a8b400-c03a-11ea-905c-6e1dab2e6d01.png">
In Safari, I cannot upload a file as it does not show any UI.

```json
    "exercises": [
        {
            "type": "file-upload",
            "id": 1158,
            "title": "UPLOAD",
            "maxScore": 1,
            "difficulty": "HARD",
            "mode": "INDIVIDUAL",
            "presentationScoreEnabled": false,
            "studentParticipations": [
                {
                    "type": "student",
                    "id": 6122,
                    "initializationState": "INITIALIZED",
                    "initializationDate": "2020-07-07T10:13:46.209034+02:00",
                    "submissions": [
                        {
                            "submissionExerciseType": "file-upload",
                            "id": 12116,
                            "submitted": false
                        }
                    ],
                    "student": {
                        "id": 2,
                        "login": "artemis_test_user_1",
                        "firstName": "ArTEMiS Test User 1",
                        "email": "krusche+testuser_1@in.tum.de",
                        "activated": false,
                        "langKey": "en",
                        "lastNotificationRead": "2020-07-05T20:50:13+02:00",
                        "name": "ArTEMiS Test User 1",
                        "participantIdentifier": "artemis_test_user_1"
                    },
                    "participantIdentifier": "artemis_test_user_1",
                    "participantName": "ArTEMiS Test User 1"
                }
            ],
            "filePattern": "pdf,png,docx,jpg,svg",
            "ended": false,
            "teamMode": false,
            "visibleToStudents": true,
            "studentAssignedTeamIdComputed": false,
            "released": true
        }
    ],
```

jpbernius commented:
Also happens in chrome and when setting release/end date:
```json
  "exercises": [
    {
      "type": "file-upload",
      "id": 1158,
      "title": "UPLOAD",
      "releaseDate": "2020-07-07T10:15:42+02:00",
      "dueDate": "2020-07-07T10:30:43+02:00",
      "maxScore": 1,
      "difficulty": "HARD",
      "mode": "INDIVIDUAL",
      "presentationScoreEnabled": false,
      "studentParticipations": [
        {
          "type": "student",
          "id": 6122,
          "initializationState": "INITIALIZED",
          "initializationDate": "2020-07-07T10:13:46.209034+02:00",
          "submissions": [
            {
              "submissionExerciseType": "file-upload",
              "id": 12116,
              "submitted": false
            }
          ],
          "student": {
            "id": 2,
            "login": "artemis_test_user_1",
            "firstName": "ArTEMiS Test User 1",
            "email": "krusche+testuser_1@in.tum.de",
            "activated": false,
            "langKey": "en",
            "lastNotificationRead": "2020-07-05T20:50:13+02:00",
            "name": "ArTEMiS Test User 1",
            "participantIdentifier": "artemis_test_user_1"
          },
          "participantIdentifier": "artemis_test_user_1",
          "participantName": "ArTEMiS Test User 1"
        }
      ],
      "filePattern": "pdf,png,docx,jpg,svg",
      "ended": false,
      "teamMode": false,
      "visibleToStudents": true,
      "studentAssignedTeamIdComputed": false,
      "released": true
    }
  ],
```

TobiasPr commented:
I think I found something unrelated: When clicking the refreshing button while being offline, it never stops. Also when I get online again after that, all the buttons are still locked - see screenshots
![image](https://user-images.githubusercontent.com/22955066/86769360-576e0980-c04f-11ea-8c45-255fa23c8043.png)
![image](https://user-images.githubusercontent.com/22955066/86769380-5dfc8100-c04f-11ea-8966-f4fabdf515c4.png)
![image](https://user-images.githubusercontent.com/22955066/86769438-740a4180-c04f-11ea-92ca-687c640e48b5.png)

anditurdiu commented:
> I think I found something unrelated: When clicking the refreshing button while being offline, it never stops. Also when I get online again after that, all the buttons are still locked, and will never become unlocked (not sure if this only occurs, when the file was not correctly initiated, due to beeing offline) - see screenshots
> ![image](https://user-images.githubusercontent.com/22955066/86769360-576e0980-c04f-11ea-8c45-255fa23c8043.png)
> ![image](https://user-images.githubusercontent.com/22955066/86769380-5dfc8100-c04f-11ea-8966-f4fabdf515c4.png)
> ![image](https://user-images.githubusercontent.com/22955066/86769438-740a4180-c04f-11ea-92ca-687c640e48b5.png)

This is a similar issue as the one I fixed for saving exercises. But since it is unrelated I think this shouldn't be a blocker. Nonetheless I added a timer to unlock the refreshing state, similar to the way I implemented it for unlocking the save state

pzdr7 commented:
> Save, Submit, Refresh files not available in exam mode after being offline

This now seems to work as expected, although I believe @TobiasPr's comment poses a similar issue here.

Something I've noticed: 

I consistently seem to get the error message about saving failing if I make changes in the code editor and then switch to a different exercise:

1. Create and prepare an exam with 2 programming exercises
2. Participate in the exam
3. Change something in the code editor of the first exercise
4. Immediately switch to the other exercise, triggering a save
5. Wait a bit, an error message should appear and leave the changes unsaved:

![image](https://user-images.githubusercontent.com/38403547/86771162-0ca1c100-c052-11ea-8876-9223f99dba0f.png)

TobiasPr commented:
Another thing I noticed: Clicking on the save button does not reset the 30s save timer - I don't know how we specified that though - is that intended?

anditurdiu commented:
> Another thing I noticed: Clicking on the save button does not reset the 30s save timer - I don't know how we specified that though - is that intended?

There was no specified behaviour for this.

kloessst commented:
@TobiasPr The Save button doesn't know about the timer
